File: javascript/pt3/js/mapObjects/gameMap.js
var GameMap = function(width, height) {
	this.width = width;
	this.height = height;
	
	this.tiles = this.initTiles();
}

GameMap.prototype.initTiles = function() {
	var tiles = this.create2dArray(this.width);
	
	for (var i = 0; i < this.width; i++) {
		for (var j = 0; j < this.height; j++) {
			tiles[i][j] = new Tile(true);
		}
	}
	
	return tiles;
}

GameMap.prototype.makeMap = function(maxRooms, roomMinSize, roomMaxSize, mapWidth, mapHeight, player) {
	var rooms = [];
	var numRooms = 0;
	
	for (var i = 0; i < maxRooms; i++) {
		var width = this.getRandomInt(roomMinSize, roomMaxSize);
		var height = this.getRandomInt(roomMinSize, roomMaxSize);
		
		var x = this.getRandomInt(0, mapWidth - width - 1);
		var y = this.getRandomInt(0, mapHeight - height - 1);
		
		var newRoom = new Rect(x, y, width, height);
		
		var intersects = false;
		for (var room of rooms) {
			if (room.intersects(newRoom)) {
				intersects = true;
				break;
			}
		}
		
		if (!intersects) {
			var newX = newRoom.getCenterX();
			var newY = newRoom.getCenterY();
			
			this.createRoom(newRoom);
			
			if (numRooms == 0) {
				// this is the first room, where we'll put the player to start atan
				player.x = newX;
				player.y = newY;
			} else {
				// all rooms after the first
				// connect this room to the previous with a tunnel
				
				var prevRoom = rooms[numRooms - 1];
				var prevX = prevRoom.getCenterX();
				var prevY = prevRoom.getCenterY();

				if (this.getRandomInt(0, 1) == 1) {
					// horizontal first, then vertical
					this.createHorizontalTunnel(prevX, newX, prevY);
					this.createVerticalTunnel(prevY, newY, newX);
				} else {
					// vertical first, then horizontal
					this.createVerticalTunnel(prevY, newY, prevX);
					this.createHorizontalTunnel(prevX, newX, newY);
				}
			}
			
			rooms.push(newRoom);
			numRooms += 1;
		}
	}
}

GameMap.prototype.createRoom = function(roomRect) {
	for (var x = roomRect.x1 + 1; x < roomRect.x2; x++) {
		for (var y = roomRect.y1 + 1; y < roomRect.y2; y++) {
			this.tiles[x][y].setBlocked(false);
		}
	}
}

GameMap.prototype.createHorizontalTunnel = function(x1, x2, y) {
	for (var x = Math.min(x1, x2); x < Math.max(x1, x2) + 1; x++) {
		this.tiles[x][y].setBlocked(false);
	}
}

GameMap.prototype.createVerticalTunnel = function(y1, y2, x) {
	for (var y = Math.min(y1, y2); y < Math.max(y1, y2) + 1; y++) {
		this.tiles[x][y].setBlocked(false);
	}
}

GameMap.prototype.isBlocked = function(x, y) {
	var blocked = false;

	if (x < 0 || y < 0 || x >= this.width || y >= this.height) {
		blocked = true;
	} else {
		blocked = this.tiles[x][y].blocked;
	}
	
	return blocked;
}

// TODO: Move this to generic util?
GameMap.prototype.create2dArray = function(rows) {
	var array = [];
	
	for (var i = 0; i < rows; i++) {
		array[i] = [];
	}
	
	return array;
}

GameMap.prototype.getRandomInt = function(min, max) {
	return Math.floor(Math.random() * (max - min + 1)) + min;
}

File: python/pt4/map_objects/game_map.py
from random import randint

from map_objects.rectangle import Rect
from map_objects.tile import Tile

class GameMap:
	def __init__(self, width, height):
		self.width = width
		self.height = height
		self.tiles = self.initialize_tiles()
		
	def initialize_tiles(self):
		tiles = [[Tile(True) for y in range(self.height)] for x in range(self.width)]
		
		return tiles
		
	def make_map(self, max_rooms, room_min_size, room_max_size, map_width, map_height, player):
		rooms = []
		num_rooms = 0
		
		for r in range(max_rooms):
			# Random width and height
			w = randint(room_min_size, room_max_size)
			h = randint(room_min_size, room_max_size)
			# Random position without going out of the map boundaries
			x = randint(0, map_width - w - 1)
			y = randint(0, map_height - h - 1)
			
			new_room = Rect(x, y, w, h)
			
			# run through the other rooms and see if the intersect with this one
			for other_room in rooms:
				if new_room.intersect(other_room):
					break
			else:
				# No intersections meaning valid room
				
				# "paint" it to the map's tiles
				self.create_room(new_room)
				
				# center coordinates of new room, will be useful later
				(new_x, new_y) = new_room.center()
				
				if num_rooms == 0:
					# this is the first room, where the player starts at
					player.x = new_x
					player.y = new_y
				else:
					# all rooms after the first:
					# connect it to the previous room with a tunnel
					
					# center coordinates of previous room
					(prev_x, prev_y) = rooms[num_rooms - 1].center()
					
					# flip a coin (Random number that is 0 or 1)
					if randint(0, 1) == 1:
						# first move horizontally, then vertically
						self.create_h_tunnel(prev_x, new_x, prev_y)
						self.create_v_tunnel(prev_y, new_y, new_x)
					else:
						# first move vertically, then horizontally
						self.create_v_tunnel(prev_y, new_y, prev_x)
						self.create_h_tunnel(prev_x, new_x, new_y)
						
				# finally, append the new room to the list
				rooms.append(new_room)
				num_rooms += 1
				
	def create_room(self, room):
		# Go through the tiles in the rectangle and make them passable
		for x in range(room.x1 + 1, room.x2):
			for y in range(room.y1 + 1, room.y2):
				self.tiles[x][y].blocked = False
				self.tiles[x][y].block_sight = False
	
	def create_h_tunnel(self, x1, x2, y):
		for x in range(min(x1, x2), max(x1, x2) + 1):
			self.tiles[x][y].blocked = False
			self.tiles[x][y].block_sight = False
			
	def create_v_tunnel(self, y1, y2, x):
		for y in range(min(y1, y2), max(y1, y2) + 1):
			self.tiles[x][y].blocked = False
			self.tiles[x][y].block_sight = False
	
	def is_blocked(self, x, y):
		if self.tiles[x][y].blocked:
			return True
		
		return False